src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,160 @@
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+            }
+
+            currentNode.PackageSources.Add(packageSourceKey);
+        }
+
+        public ConfigNameSpaceLookup Find(string term)
+        {
+            if (string.IsNullOrWhiteSpace(term))
+            {
+                throw new ArgumentNullException(nameof(term));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            term = term.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+            SearchNode currentNode = _root;
+            int i = 0;
+
+            for (; i < term.Length; i++)
+            {
+                char c = term[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    return currentNode.IsGlobbing ? new ConfigNameSpaceLookup(true, currentNode.PackageSources) : new ConfigNameSpaceLookup(false, currentNode.PackageSources);
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (i == term.Length && !currentNode.IsValueNode) // full 'term' already matched, but still not at value Node. That means we need to go backtrace try to find better matching sources.
+            {
+                while (currentNode != null && !currentNode.IsValueNode)
+                {
+                    currentNode = currentNode.Parent;
+                }
+            }
+
+            return currentNode == null ? new ConfigNameSpaceLookup(false, null)
+                                        : new ConfigNameSpaceLookup(true, currentNode.PackageSources);
+        }
+
+        public static SearchTree GetSearchTree(ISettings settings, ILogger logger)
+        {
+            SearchTree nameSpaceLookup = null;
+
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            PackageNamespacesConfiguration configuration = PackageNamespacesConfiguration.GetPackageNamespacesConfiguration(settings);
+            var packageSourceSections = new List<PackageSourceSection>();
+
+            foreach (var packageSourceKey in configuration.Namespaces.Keys)
+            {
+                string[] nugetNamespaces = configuration.Namespaces[packageSourceKey].ToArray();
+                packageSourceSections.Add(new PackageSourceSection(packageSourceKey, nugetNamespaces));
+            }
+
+            if (packageSourceSections.Any())
+            {
+                nameSpaceLookup = new SearchTree(packageSourceSections);
+                logger?.LogDebug(string.Format(CultureInfo.CurrentCulture, Resources.PackageNamespaceFound));
```

Comment:
Is this trying to articulate that namespaces will be used? 

In that case, I'd avoid logging this message from a shared type. 

It should be the scenario, such as restore, install etc. that decides what messages we log. Things hidden in shared code can be harder to track.

Reply:
Remove this one.